### ISSUE TITLE
Errors reported when using AnyStr.join(Iterable[AnyStr])

### ISSUE THREAD

**Issue by tpoliaw** [microsoft/pyright#5556]
**Describe the bug**
Calling `AnyStr.join(Iterable[AnyStr])` is flagged as an error where the types should always match.

**To Reproduce**
```python
from typing import AnyStr, Iterable

def joiner(delim: AnyStr, objs: Iterable[AnyStr]) -> AnyStr:
    return delim.join(objs)
```
Running `pyright type_demo.py` gives the following error
```
/tmp/type_demo.py
  /tmp/type_demo.py:4:23 - error: Argument of type "Iterable[AnyStr@joiner]" cannot be assigned to parameter "__iterable" of type "Iterable[str]" in function "join"
    "Iterable[AnyStr@joiner]" is incompatible with "Iterable[str]"
      Type parameter "_T_co@Iterable" is covariant, but "AnyStr@joiner" is not a subtype of "str" (reportGeneralTypeIssues)
  /tmp/type_demo.py:4:23 - error: Argument of type "Iterable[AnyStr@joiner]" cannot be assigned to parameter "__iterable_of_bytes" of type "Iterable[ReadableBuffer]" in function "join"
    "Iterable[AnyStr@joiner]" is incompatible with "Iterable[str]"
      Type parameter "_T_co@Iterable" is covariant, but "AnyStr@joiner" is not a subtype of "str" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

**Expected behavior**
This should be accepted as a valid call. `str.join(Iterable[str])` and `bytes.join(Iterable[bytes])` both pass as expected.

**Code or Screenshots**
see above

**VS Code extension or command-line**
Running from the command line. Version 1.1.318

**Additional context**
I raised this as an [issue with typeshed](https://github.com/python/typeshed/issues/10495) and they suggested this was an issue on pyright's side.

**Comment by erictraut** [comment]
`AnyStr` is a constrained TypeVar. Constrained TypeVars are unique to the Python type system, motivated primarily by `AnyStr`. They're a very odd construct in a type system, and they're difficult for a type checker to handle correctly in all cases. There are many edge cases that pyright cannot handle — and a different set of edge cases that mypy cannot handle. For additional details, refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=constrained-type-variables).

I'll spend some time investigating whether there's a way to handle this particular case. Just to set expectations, the answer might be "pyright can't support this". In this case, you'll need to use a `# type: ignore` or some other workaround if you don't want to change your code.

**Comment by tpoliaw** [comment]
Ok, no problem - we can ignore it for now.

**Comment by erictraut** [comment]
I found a way to make this case work. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.319, which I just published. It will also be included in a future release of pylance.
